### ISSUE TITLE
Write Float bug

### ISSUE THREAD

**Issue by fedjabosnic** [fedjabosnic/hotfix#44]
A number like `0.00012345` incorrectly gets written as `0.12345___` _(underscore represents blank bytes)_

@GeorgeF0 any quick thoughts on how to fix?

**Comment by GeorgeF0** [comment]
```csharp
        /// <summary>
        /// Writes a long to the given byte array. 
        /// The number is written before the given position, so that the last digit is at that position.
        /// </summary>
        /// <param name="buffer">The buffer to write to.</param>
        /// <param name="position">The position that marks the end of the number.</param>
        /// <param name="value">The number to write.</param>
        /// <returns>The number of characters written</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int WriteLongBackwards(this byte[] buffer, int position, long value)
        {
            if (value == 0)
            {
                buffer[position] = (byte)'0';
                return 1;
            }

            var initial = position;
            var negative = false;

            if (value < 0)
            {
                negative = true;
                value = -value; // Note: Code won't work for long.MinValue
            }

            while (value > 0)
            {
                var n = '0' + (char)(value % 10);
                buffer[position--] = (byte)n;
                value /= 10;
            }

            if (negative) buffer[position--] = (byte)'-';

            return initial - position;
        }

        /// <summary>
        /// Writes a double to the given byte array. 
        /// The number is rounded to 6 decimal places.
        /// </summary>
        /// <param name="buffer">The buffer to write to.</param>
        /// <param name="position">The position to start writing.</param>
        /// <param name="value">The number to write.</param>
        /// <returns>The number of characters written.</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int WriteFloat(this byte[] buffer, int position, double value)
        {
            var start = position;
            var integer = (long)value;
            var fraction = (long)Math.Round(Math.Abs(value) * 1000000L) % 1000000L;

            position += buffer.WriteLong(position, integer);
            buffer[position++] = (byte)'.';
            var written = buffer.WriteLongBackwards(position + 5, fraction);

            for (var i = position; i < position + 6 - written; i++)
            {
                buffer[i] = (byte)'0';
            }

            return position + 6 - start;
        }
```

something like this?

The idea is to write the fractional part backwards, since we know it's 6 dp (we "know" as in, that's what we decided to use), and fill in any leading zeroes with the '0' character. A consequence of that is that numbers with a 0 fractional part are printed with 6 zeroes rather than 1.

**Comment by GeorgeF0** [comment]
Come to think of it, if it's only 6dp, `WriteIntBackwards` could also work. So something like this should be enough

```csharp
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int WriteFloat(this byte[] buffer, int position, double value)
        {
            var start = position;
            var integer = (long)value;
            var fraction = (int)Math.Round(Math.Abs(value) * 1000000L) % 1000000;

            position += buffer.WriteLong(position, integer);
            buffer[position++] = (byte)'.';
            var written = buffer.WriteIntBackwards(position + 5, fraction);

            for (var i = position; i < position + 6 - written; i++)
            {
                buffer[i] = (byte)'0';
            }

            return position + 6 - start;
        }
```

**Comment by fedjabosnic** [comment]
Good ideas - I went with the first one (using longs) but I made the decimal places a parameter as it seems like we would want more control long-term.

See pull request #45 - it seems to work and performance isn't affected much...

**Comment by fedjabosnic** [comment]
Fixed in #45
